api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -300,101 +299,88 @@ private DataSetRequest generateDataSetRequestFromPreviewRequest(
-                              valuePreviewList.get(columnNumber).addQueryValueItem("");
-                            }
-                          });
-                });
-        queryResponse
-            .getSchema()
-            .getFields()
-            .forEach(
-                fields -> {
-                  DataSetPreviewValueList previewValue =
-                      valuePreviewList.stream()
-                          .filter(preview -> preview.getValue().equalsIgnoreCase(fields.getName()))
-                          .findFirst()
-                          .get();
-                  if (fields.getType() == LegacySQLTypeName.TIMESTAMP) {
-                    List<String> queryValues = new ArrayList<String>();
-                    DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
-                    previewValue
-                        .getQueryValue()
-                        .forEach(
-                            value -> {
-                              try {
-                                Double fieldValue = Double.parseDouble(value);
-                                queryValues.add(
-                                    dateFormat.format(new Date(fieldValue.longValue())));
-                              } catch (NumberFormatException ex) {
-                                queryValues.add("");
-                              }
-                            });
-                    previewValue.setQueryValue(queryValues);
-                  }
-                });
-        break;
-      } catch (Exception ex) {
-        if ((ex.getCause() != null
-            && ex.getCause().getMessage() != null
-            && ex.getCause().getMessage().contains("Read timed out"))) {
-          rowsRequested = (rowsRequested / 2);
-          if (rowsRequested == 0) {
-            throw new GatewayTimeoutException(
-                "Timeout while querying the CDR to pull preview information.");
-          }
-          retry++;
-        } else {
-          throw ex;
-        }
+    TableResult queryResponse;
+    try {
+      String query = originalQuery.concat(" LIMIT " + NO_OF_PREVIEW_ROWS);
+
+      queryJobConfiguration = queryJobConfiguration.toBuilder().setQuery(query).build();
+
+      /* Google appengine has a 60 second timeout, we want to make sure this endpoint completes
+       * before that limit is exceeded, or we get a 500 error with the following type:
+       * com.google.apphosting.runtime.HardDeadlineExceededError
+       * See https://cloud.google.com/appengine/articles/deadlineexceedederrors for details
+       */
+      queryResponse =
+          bigQueryService.executeQuery(
+              bigQueryService.filterBigQueryConfig(queryJobConfiguration),
+              APP_ENGINE_HARD_TIMEOUT_MSEC_MINUS_FIVE_SEC);
+    } catch (Exception ex) {
+      if ((ex.getCause() != null
+          && ex.getCause().getMessage() != null
+          && ex.getCause().getMessage().contains("Read timed out"))) {
```

Comment:
Does BigQuery not document a more specific exception in this situation? Looks like we can catch [BigQueryException](https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/bigquery/BigQueryException.html#getError--) and then call `getError` on that to get a `BigQueryError`, which exposes [getReason](https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/bigquery/BigQueryError.html#getReason--), which is a "short error code", which might be more stable than using the message.

Reply:
Let me see what I can do.

Reply:
For posterity, I poked around at this, but we do some abstraction over the BigQueryException, which means that isn't what we get back from the `executeQuery` method. I tried to see if I could consistently get an error from bigquery in order to resolve this, but I couldn't. Talked offline with Jay, who said if I spent too long I can just punt on this for now, because it is an existing problem in the codebase.